mllib/src/main/scala/org/apache/spark/ml/feature/Instance.scala

```diff
@@ -114,6 +133,74 @@ private[spark] object InstanceBlock {
   def blokify(instances: RDD[Instance], blockSize: Int): RDD[InstanceBlock] = {
     instances.mapPartitions(_.grouped(blockSize).map(InstanceBlock.fromInstances))
   }
+
+  def blokifyWithMaxMemUsage(
+      iterator: Iterator[Instance],
+      maxMemUsage: Long): Iterator[InstanceBlock] = {
+    require(maxMemUsage > 0)
+
+    var numCols = -1L
+    val buff = mutable.ArrayBuilder.make[Instance]
+    var buffCnt = 0L
+    var buffNnz = 0L
+    var buffUnitWeight = true
+
+    iterator.flatMap { instance =>
+      if (numCols < 0L) numCols = instance.features.size
+      require(numCols == instance.features.size)
+      val nnz = instance.features.numNonzeros
+      buff += instance
+      buffCnt += 1L
+      buffNnz += nnz
+      buffUnitWeight &&= (instance.weight == 1)
+
+      if (getBlockMemUsage(numCols, buffCnt, buffNnz, buffUnitWeight) >= maxMemUsage) {
+        val block = InstanceBlock.fromInstances(buff.result())
+        buff.clear()
+        buffCnt = 0L
+        buffNnz = 0L
+        buffUnitWeight = true
+        Iterator.single(block)
+      } else Iterator.empty
+    } ++ {
+      if (buffCnt > 0) {
+        val block = InstanceBlock.fromInstances(buff.result())
+        Iterator.single(block)
+      } else Iterator.empty
+    }
+  }
+
+  def blokifyWithMaxMemUsage(
+      instances: RDD[Instance],
+      maxMemUsage: Long): RDD[InstanceBlock] = {
+    require(maxMemUsage > 0)
+    instances.mapPartitions(iter => blokifyWithMaxMemUsage(iter, maxMemUsage))
+  }
+
+
+  /**
+   * Suggested value for BlockSizeInMB, based on performance tests of BLAS operation.
+   *
+   * @param dim size of vector.
+   * @param avgNNZ average nnz of vectors.
+   * @param blasLevel level of BLAS operation.
+   */
+  def inferBlockSizeInMB(
+      dim: Int,
+      avgNNZ: Double,
+      blasLevel: Int = 2): Double = {
+    if (dim <= avgNNZ * 3) {
+      // When the dataset is relatively dense, Spark will use netlib-java for optimised numerical
+      // processing, which will try to use nativeBLAS implementations (like OpenBLAS, Intel MKL),
+      // and fallback to the Java implementation (f2jBLAS) if necessary.
+      // The suggested value for dense cases is 0.25.
+      0.25
```

Comment:
We need more comments to explain how this default value is picked. So this is roughly 180x180 double, which seems quite small to me.

Reply:
We may also change it to 1.0 for dence cases (to use 1.0 as the default value for all cases), the speedup at 1.0MB is only a little lower than that at 0.25MB.


There was [another performance test](https://issues.apache.org/jira/browse/SPARK-31714) on the implements of prediction in training, which maybe worthwhile to refer to:

```
test("performance: gemv vs foreachNonZero(std)") {
  for (numRows <- Seq(16, 64, 256, 1024, 4096); numCols <- Seq(16, 64, 256, 1024, 4096)) {
    val rng = new Random(123)
    val matrix = Matrices.dense(numRows, numCols,
      Array.fill(numRows * numCols)(rng.nextDouble)).toDense
    val vectors = matrix.rowIter.toArray
    val coefVec = Vectors.dense(Array.fill(numCols)(rng.nextDouble))
    val coefArr = coefVec.toArray
    val stdVec = Vectors.dense(Array.fill(numCols)(rng.nextDouble))
    val stdArr = stdVec.toArray

    val start1 = System.nanoTime
    Seq.range(0, 100).foreach { _ => matrix.multiply(coefVec) }
    val dur1 = System.nanoTime - start1

    val start2 = System.nanoTime
    Seq.range(0, 100).foreach { _ =>
      vectors.map { vector =>
        var sum = 0.0
        vector.foreachNonZero { (i, v) =>
          val std = stdArr(i)
          if (std != 0) sum += coefArr(i) * v
        }
        sum
      }
    }
    val dur2 = System.nanoTime - start2

    println(s"numRows=$numRows, numCols=$numCols, gemv: $dur1, foreachNonZero(std): $dur2, " +
      s"foreachNonZero(std)/gemv: ${dur2.toDouble / dur1}")
  }
}
```

output:
```
numRows=16, numCols=16, gemv: 543897, foreachNonZero(std): 4683864, foreachNonZero(std)/gemv: 8.611674636925741
numRows=16, numCols=64, gemv: 274878, foreachNonZero(std): 2996356, foreachNonZero(std)/gemv: 10.90067593623353
numRows=16, numCols=256, gemv: 771816, foreachNonZero(std): 9081260, foreachNonZero(std)/gemv: 11.76609450957223
numRows=16, numCols=1024, gemv: 1537698, foreachNonZero(std): 23386693, foreachNonZero(std)/gemv: 15.208898626388276
numRows=16, numCols=4096, gemv: 5577804, foreachNonZero(std): 87389503, foreachNonZero(std)/gemv: 15.667367121541023
numRows=64, numCols=16, gemv: 173518, foreachNonZero(std): 1384669, foreachNonZero(std)/gemv: 7.979973259258405
numRows=64, numCols=64, gemv: 313941, foreachNonZero(std): 4403461, foreachNonZero(std)/gemv: 14.026396679630887
numRows=64, numCols=256, gemv: 981895, foreachNonZero(std): 19443231, foreachNonZero(std)/gemv: 19.801741530408037
numRows=64, numCols=1024, gemv: 3908960, foreachNonZero(std): 88985415, foreachNonZero(std)/gemv: 22.764473159101144
numRows=64, numCols=4096, gemv: 16075758, foreachNonZero(std): 366740675, foreachNonZero(std)/gemv: 22.81327418588909
numRows=256, numCols=16, gemv: 329479, foreachNonZero(std): 5341171, foreachNonZero(std)/gemv: 16.210960334346044
numRows=256, numCols=64, gemv: 948949, foreachNonZero(std): 17126600, foreachNonZero(std)/gemv: 18.047966750584067
numRows=256, numCols=256, gemv: 3947789, foreachNonZero(std): 81207071, foreachNonZero(std)/gemv: 20.570266293360664
numRows=256, numCols=1024, gemv: 14635992, foreachNonZero(std): 350779742, foreachNonZero(std)/gemv: 23.96692632791819
numRows=256, numCols=4096, gemv: 71265609, foreachNonZero(std): 1423000813, foreachNonZero(std)/gemv: 19.967566866649523
numRows=1024, numCols=16, gemv: 916645, foreachNonZero(std): 19432942, foreachNonZero(std)/gemv: 21.200074183571612
numRows=1024, numCols=64, gemv: 3479825, foreachNonZero(std): 66857430, foreachNonZero(std)/gemv: 19.212871336920678
numRows=1024, numCols=256, gemv: 13680423, foreachNonZero(std): 312189763, foreachNonZero(std)/gemv: 22.82018348409256
numRows=1024, numCols=1024, gemv: 68880268, foreachNonZero(std): 1401019163, foreachNonZero(std)/gemv: 20.339920323771096
numRows=1024, numCols=4096, gemv: 293455450, foreachNonZero(std): 5744847994, foreachNonZero(std)/gemv: 19.576559215376644
numRows=4096, numCols=16, gemv: 3714086, foreachNonZero(std): 82488401, foreachNonZero(std)/gemv: 22.20960984748334
numRows=4096, numCols=64, gemv: 14273712, foreachNonZero(std): 279946980, foreachNonZero(std)/gemv: 19.612766461870606
numRows=4096, numCols=256, gemv: 70000687, foreachNonZero(std): 1311574476, foreachNonZero(std)/gemv: 18.73659434228124
numRows=4096, numCols=1024, gemv: 289944283, foreachNonZero(std): 5695483201, foreachNonZero(std)/gemv: 19.643371278336257
numRows=4096, numCols=4096, gemv: 1169773295, foreachNonZero(std): 23019445987, foreachNonZero(std)/gemv: 19.678553173843827
- performance: gemv vs foreachNonZero(std)
```

new implements are based on `gemv`, while implements in branch-3.0 are beased on `foreachNonZero(std)`.

I think that a blockSize larger than 64X256 may be acceptable.